packages/contracts/contracts/L1/rollup/CanonicalTransactionChain.sol

```diff
@@ -61,12 +67,70 @@ contract CanonicalTransactionChain is ICanonicalTransactionChain, Lib_AddressRes
         Lib_AddressResolver(_libAddressManager)
     {
         maxTransactionGasLimit = _maxTransactionGasLimit;
-        L2_GAS_DISCOUNT_DIVISOR = _l2GasDiscountDivisor;
-        ENQUEUE_GAS_COST  = _enqueueGasCost;
-        ENQUEUE_L2_GAS_PREPAID = _l2GasDiscountDivisor * _enqueueGasCost;
+        l2GasDiscountDivisor = _l2GasDiscountDivisor;
+        enqueueGasCost  = _enqueueGasCost;
+        enqueueL2GasPrepaid = _l2GasDiscountDivisor * _enqueueGasCost;
     }
 
 
+
+    /**********************
+     * Function Modifiers *
+     **********************/
+
+    /**
+     * Modifier to enforce that, if configured, only the OVM_Sequencer contract may
+     * successfully call a method.
+     */
+    modifier onlySequencer() {
+        require(
+            msg.sender == resolve("OVM_Sequencer"),
+            "Only callable by the Sequencer."
+        );
+        _;
+    }
+
+    /*******************************
+     * Authorized Setter Functions *
+     *******************************/
+
+    /**
+     * Allows the Sequencer to update the gas amount which is 'prepaid' during enqueue.
+     * The value of enqueueL2GasPrepaid is immediately updated as well.
+     */
+    function setEnqueueGasCost(uint256 _enqueueGasCost)
+        external
+        onlySequencer
+    {
+        enqueueGasCost = _enqueueGasCost;
+        enqueueL2GasPrepaid = l2GasDiscountDivisor * _enqueueGasCost;
+
+        emit L2GasParamsUpdated(
+            l2GasDiscountDivisor,
+            enqueueGasCost,
+            enqueueL2GasPrepaid
+        );
+    }
+
+    /**
+     * Allows the Sequencer to update the L2 Gas Discount Divisor, which is defined as the ratio
+     * of the cost of gas on L1 to L2.
+     * The value of enqueueL2GasPrepaid is immediately updated as well.
+     */
+    function setGasDivisor(uint256 _l2GasDiscountDivisor)
+        external
+        onlySequencer
+    {
+        l2GasDiscountDivisor = _l2GasDiscountDivisor;
+        enqueueL2GasPrepaid = _l2GasDiscountDivisor * enqueueGasCost;
```

Comment:
The reason why we multiply by `_l2GasDiscountDivisor` took me a long time to understand. This is my understanding:

We supply a new `l2GasDiscountDivisor` in this function. This changes how much gas we "pre-pay" for because we always prepay for the cost to enqueue (denominated in L1 gas), which to convert to l2 gas we have to multiply by the `_l2GasDiscountDivisor`.

In check that is performed `if(_gasLimit > enqueueL2GasPrepaid) {`, `_gasLimit`'s unit is L2 gas. That's why when we compare `enqueueL2GasPrepaid` we need to make sure that we are comparing it to the *L2 gas* which is prepaid. And so the above conversion where we multiply by the `_l2GasDiscountDivisor` is super key

Reply:
Yes, well put!

Reply:
I will add some comments where the gas burn occurs to expand on it further.

Reply:
I expanded on the calculation here: 
9f5a2b1d5c1bc8ffe5a6a4a36dc12f6b5a2f2e9a

Reply:
Dope!